### ISSUE TITLE
false import-errors with pip >= 21.3

### ISSUE THREAD

**Issue by belm0** [pylint-dev/pylint#7306]
### Bug description

summary:
  * pylint incorrectly reports `import-error` with recent pip versions (>= 21.3)
  * the environment uses `pip install -e .`, on a project with `setup.py` and `pyproject.toml`
  * confirmed normal runtime doesn't have import errors, and pylint has no errors with pip < 21.3
  * import of pylint plugins local to the project is also affected by the bug

The following change from pip 21.3.0 may be relevant:
> Support editable installs for projects that have a pyproject.toml and use a build backend that supports PEP 660 (https://github.com/pypa/pip/issues/8212).

create project & env:
```shell
echo 'import setuptools; setuptools.setup(name="my_project", packages=["foo"])' > setup.py
touch pyproject.toml  # toml must exist to reproduce the problem
mkdir foo scripts
touch foo/__init__.py
echo 'import foo' > scripts/bar.py

python -m venv .venv
source .venv/bin/activate
pip install --upgrade pip pylint
pip install -e .
```

### Command used

```shell
pylint scripts/bar.py
```

### Pylint output

```shell
************* Module bar
scripts/bar.py:2:0: E0401: Unable to import 'foo' (import-error)
```

### Expected behavior

no import-error, as with earlier pip versions

### Pylint version

```shell
pylint 2.14.5
astroid 2.11.7
Python 3.8.12 (default, Sep 23 2021, 08:42:37)
[Clang 12.0.0 (clang-1200.0.32.29)]
```

### Additional dependencies

pip => 21.3

**Comment by jooola** [comment]
I am having a similar error, but on my side it happens when using setuptools >=64.0.0 which introduced https://togithub.com/pypa/setuptools/pull/3488. Have you tried pinning the setuptools version below 64.0.0 for the build backend ?

**Comment by jooola** [comment]
Here is a minimal reproducible repository https://github.com/jooola/pylint-import-error-bug

**Comment by DanielNoord** [comment]
@jooola Can you confirm that for the reproducing repository doing `pip install setuptools==63` and then reinstalling the editable packages fixes the issue? I'm having trouble reproducing the successful behaviour on lower setuptools versions locally.

**Comment by jooola** [comment]
@DanielNoord I was was able to fix my issue buy applying the patch I just pushed. Which pin the setuptools version below 64.0.0

https://github.com/jooola/pylint-import-error-bug/blob/main/before_setuptools_v64.0.0.patch

**Comment by jooola** [comment]
Ok, I can confirm this again, setuptools<63 fixes the bug.

**Comment by DanielNoord** [comment]
The issue is in the following piece of code:
https://github.com/PyCQA/astroid/blob/d033fe2f4c575d45a1706c52988dbcdf50645c5c/astroid/interpreter/_import/spec.py#L135-L145

On `setuptools<64` we successfully find the `shared_lib` package via `sys.path` and looking for its `__init__.py` file.
On `setuptools>65` this no longer works. I think this is because editable installs are no longer put on `sys.path` but I'm note 100% sure. It is definitely due to the `setuptools` changes though.

The repository with the pinning of `setuptools` reproduces the issues consistently.

I'm not sure I'll be able to get to this this week so if anybody is interested please go ahead and investigate more or propose a solution. I'm inclined to open an issue against `setuptools` to ask them how they think we should discover editable packages with the new backend.

**Comment by MiltiadisKoutsokeras** [comment]
pylint is used in our workflows and currently it breaks them with import errors from an editable installation of our projects. Is there something we can change in the usage of pylint to avoid these errors without changing the project's code or file hierarchy?

**Comment by DanielNoord** [comment]
It's a bit of a stretch but you could install `astroid` from my branch in https://github.com/PyCQA/astroid/pull/1752.

See https://stackoverflow.com/questions/20101834/pip-install-from-git-repo-branch
Or `git clone` my branch and then install locally.

Fixing stuff in our import system is difficult as much can be broken. So it takes some time for these fixes to make it to release. This branch has been working for me with the new setuptools versions though.

**Comment by jooola** [comment]
> pylint is used in our workflows and currently it breaks them with import errors from an editable installation of our projects. Is there something we can change in the usage of pylint to avoid these errors without changing the project's code or file hierarchy?

You can either pin the setuptools version ([see the bug reproduction repository](https://github.com/jooola/pylint-import-error-bug/blob/main/before_setuptools_v64.0.0.patch)) or add the [legacy editable option](https://setuptools.pypa.io/en/latest/userguide/development_mode.html#legacy-behavior) when installing.

**Comment by MiltiadisKoutsokeras** [comment]
Thank you for your recommendations but the issue is reproduced with these versions:

Python 3.9.2
pip 22.2.2
setuptools 44.1.1

So setuptools is less than 64.

**Comment by DanielNoord** [comment]
Even then, installing my local branch should fix this as it add support for custom import hooks.

Also note that it's quite easy to have `pip` use a different version of setup tools than what's showing up with `pip list`. I had similar issues while working on the fix.

**Comment by MiltiadisKoutsokeras** [comment]
The only safe and non-intrusive fix for now is `export SETUPTOOLS_ENABLE_FEATURES="legacy-editable"` in the Bash script that prepares the Python venv and installs the editable version. Since this is temporary and going to be eventually deprecated I hope there will be a more efficient solution to support the new editable install format.

**Comment by Pierre-Sassoulas** [comment]
It seems it might be a bug caused by https://github.com/pypa/setuptools/issues/3499 that affects setuptools ``64.0.0`` but was fixed in ``64.0.1`` ?

**Comment by niander** [comment]
Ok, so I was able to confirm that `setuptools<63.0.0` seems to fix the issue. @DanielNoord I think I know why sometimes it is hard to know which "setuptools" pip is using. I found out that doesn't matter which version I have installed in my venv. Because I have `pyproject.toml` (I also have` setup.cfg` and `setup.py`), the newer versions of `pip` seems to always load `pyproject.toml`. And, because I added to the file `requires = ["setuptools>=61.0.0", "wheel"]`, pip seems to install the latest "setuptools" and don't use the one already installed.

I changed my `pyproject.toml` to `requires = ["setuptools>=61.0.0,<63.0.0", "wheel"]` and that itself fixed the issue.
Also, adding `-v` to pip installation shows which build packages pip is installed (or loading) temporarily to install the package.

**Comment by ssbarnea** [comment]
I see that this bug happens even with the latest `setuptools==67.2.0` and I do not understand why we still consider setuptools being the cause when python imports do really work, is just pylint that thinks that they do not.

Any pinning of setuptools would produce other problems, even if done only for development.

**Comment by DanielNoord** [comment]
I don't understand your comment. There is an incompatibility with `setuptools` and `pylint` above a certain version.
Nobody is "blaming" `setuptools` we are just identifying the source of the issue.

Anyway, @Pierre-Sassoulas would you mind releasing `astroid`. I think this is a fix worth bringing out and we can start migrating the `Uninferable` stuff in `pylint`.

**Comment by Pierre-Sassoulas** [comment]
Sure let's release a new version of astroid.

**Comment by noah-weingarden** [comment]
@DanielNoord Thanks for your work on https://github.com/PyCQA/astroid/pull/1752! I installed `astroid` from its `main` branch on GitHub so that I could give the fix a try. I'm not sure if I'm doing this too early or not, but you mentioned that you'd like feedback [here](https://github.com/PyCQA/astroid/pull/1752#issuecomment-1407752582), so I'll give it a shot. Let me know if I should make a new issue rather than adding it on here.

I'm finding that whereas before `pylint` reported `import-error` for modules imported from an editable install, it now correctly identifies the modules and objects available in them like classes, but not submodules. I created a minimal reproducible example with steps to reproduce in [this repo](https://github.com/noah-weingarden/pylint-mre). After running `pylint tests`, I get this output:

```console
************* Module test_everything
tests/test_everything.py:6:14: E1101: Module 'pylint_bug_mre' has no 'goodbye' member; maybe 'Goodbye'? (no-member)
tests/test_everything.py:11:12: E1101: Module 'pylint_bug_mre' has no 'hello' member; maybe 'Hello'? (no-member)
```

It seems like pylint recognizes the module being imported and the two classes that it exports, but it doesn't recognize its submodules. I confirmed that pylint reports no errors if I use a `setup.py` instead of `pyproject.toml` (thereby using old editable install method instead of the PEP 660 one).

**Comment by belm0** [comment]
Seems to be resolved in pylint 3.0

**Comment by Pierre-Sassoulas** [comment]
Let's close then, please open a new issue referencing this one if you still encounter the issue.
